lib/index.d.ts

```diff
@@ -1067,5 +1076,276 @@ declare module "react-native-firebase" {
+          fieldPath: NativeFieldPath;
+        }
+
+        interface QueryOptions {
+          endAt?: any[];
+          endBefore?: any[];
+          limit?: number;
+          offset?: number;
+          selectFields?: string[];
+          startAfter?: any[];
+          startAt?: any[];
+        }
+
+        // The JS code expects at least one of 'includeDocumentMetadataChanges'
+        // or 'includeQueryMetadataChanges' to be true. This logic is
+        // encapsulated in QueryListenOptions.
+        interface _IncludeDocumentMetadataChanges {
+          includeDocumentMetadataChanges: true;
+        }
+        interface _IncludeQueryMetadataChanges {
+          includeQueryMetadataChanges: true
+        }
+        type QueryListenOptions = _IncludeDocumentMetadataChanges | _IncludeQueryMetadataChanges | (_IncludeDocumentMetadataChanges & _IncludeQueryMetadataChanges);
+
+        type ObserverOnNext = (querySnapshot: QuerySnapshot) => void;
+        type ObserverOnError = (err: object) => void;
+        interface Observer {
+          next: ObserverOnNext;
+          error?: ObserverOnError;
+        }
+      }
+
+      class QuerySnapshot {
+        constructor(firestore: Firestore, query: Query, nativeData: QuerySnapshot.NativeData);
+        get docChanges(): DocumentChange[];
+        get docs(): DocumentSnapshot[];
+        get empty(): boolean;
+        get metadata(): Types.SnapshotMetadata;
+        get query(): Query;
+        get size(): number;
+        forEach(callback: (snapshot: DocumentSnapshot) => any);
+      }
+      namespace QuerySnapshot {
+        interface NativeData {
+          changes: Types.NativeDocumentChange[];
+          documents: Types.NativeDocumentSnapshot[];
+          metadata: Types.SnapshotMetadata;
+        }
+      }
+
+      class WriteBatch {
+        constructor(firestore: Firestore);
+        commit(): Promise<void>;
+        delete(docRef: DocumentReference): WriteBatch;
+        set(docRef: DocumentReference, data: object, options?: Types.WriteOptions): WriteBatch;
+        // multiple overrides for update() to allow strong-typed var_args
+        update(docRef: DocumentReference, obj: object): WriteBatch;
+        update(docRef: DocumentReference, key1: Types.UpdateKey, val1: any): WriteBatch;
+        update(docRef: DocumentReference, key1: Types.UpdateKey, val1: any, key2: Types.UpdateKey, val2: any): WriteBatch;
+        update(docRef: DocumentReference, key1: Types.UpdateKey, val1: any, key2: Types.UpdateKey, val2: any, key3: Types.UpdateKey, val3: any): WriteBatch;
+        update(docRef: DocumentReference, key1: Types.UpdateKey, val1: any, key2: Types.UpdateKey, val2: any, key3: Types.UpdateKey, val3: any, key4: Types.UpdateKey, val4: any): WriteBatch;
+        update(docRef: DocumentReference, key1: Types.UpdateKey, val1: any, key2: Types.UpdateKey, val2: any, key3: Types.UpdateKey, val3: any, key4: Types.UpdateKey, val4: any, key5: Types.UpdateKey, val5: any): WriteBatch;
```

Comment:
The problem with the simpler approach is that it allows the method to be called with invalid arguments:
```javascript
update(docRef, 'field1', 2, 3)
```
Here we've got a second 'key-value pair' where the key is 3 (which is not valid, since only strings and `FieldPath` may be a key), and the value is not even in the method call (which is invalid since we must have an odd number of arguments when using key-value pairs [ref](https://github.com/invertase/react-native-firebase/blob/master/lib/modules/firestore/WriteBatch.js#L75).

The method signatures that I've written are certainly unwieldy and messy looking, and limited to only 5 key-value pairs, but they clearly document what parameters the user can pass to the function. 

I can re-write this method to use the simpler syntax if you like, however as a heavy Typescript user I would much prefer to have the longer unwieldy typings if they can guarantee that I am calling the method with the correct parameters.